packages/@tinacms/schema-tools/src/validate/fields.ts

```diff
@@ -0,0 +1,225 @@
+/**
+Copyright 2021 Forestry.io Holdings, Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { z } from 'zod'
+import { TinaFieldInner } from '../types/SchemaTypes'
+import { hasDuplicates } from '../util'
+
+const TypeName = [
+  'string',
+  'boolean',
+  'number',
+  'datetime',
+  'image',
+  'object',
+  'reference',
+  'rich-text',
+] as const
+
+const typeTypeError = `type must be one of ${TypeName.join(', ')}`
+const typeRequiredError = `type is required and must be one of ${TypeName.join(
+  ', '
+)}`
+
+const nameProp = z.string({
+  required_error: 'Name must be provided',
+  invalid_type_error: 'Name must be a sting',
+})
+
+const Option = z.union(
+  [z.string(), z.object({ label: z.string(), value: z.string() })],
+  {
+    errorMap: () => {
+      return {
+        message:
+          'Invalid option array. Must be a string[] or {label: string, value: string}[]',
+      }
+    },
+  }
+)
+const TinaField = z.object({
+  name: nameProp,
+  label: z.string().optional(),
+  description: z.string().optional(),
+  required: z.boolean().optional(),
+})
+
+const FieldWithList = TinaField.extend({ list: z.boolean().optional() })
+
+// ==========
+// Scaler fields
+// ==========
+const TinaScalerBase = FieldWithList.extend({
+  options: z.array(Option).optional(),
+})
+const StringField = TinaScalerBase.extend({
+  type: z.literal('string', {
+    invalid_type_error: typeTypeError,
+    required_error: typeRequiredError,
+  }),
+})
+const BooleanField = TinaScalerBase.extend({
+  type: z.literal('boolean' as const, {
+    invalid_type_error: typeTypeError,
+    required_error: typeRequiredError,
+  }),
+})
+const NumberField = TinaScalerBase.extend({
+  type: z.literal('number' as const, {
+    invalid_type_error: typeTypeError,
+    required_error: typeRequiredError,
+  }),
+})
+const ImageField = TinaScalerBase.extend({
+  type: z.literal('image' as const, {
+    invalid_type_error: typeTypeError,
+    required_error: typeRequiredError,
+  }),
+})
+
+const DateTimeField = TinaScalerBase.extend({
+  type: z.literal('datetime' as const, {
+    invalid_type_error: typeTypeError,
+    required_error: typeRequiredError,
+  }),
+  dateFormat: z.string().optional(),
+  timeFormat: z.string().optional(),
+})
+
+// ==========
+// Non Scaler fields
+// ==========
+const ReferenceField = FieldWithList.extend({
+  type: z.literal('reference' as const, {
+    invalid_type_error: typeTypeError,
+    required_error: typeRequiredError,
+  }),
```

Comment:
Out of scope of this ticket but - `type: 'reference'` should validate if the `collections` prop is actually referencing collections. This was something we could do in Yup, and seems a bit clunky in Zod, but maybe can be done with something like [this](https://github.com/colinhacks/zod/issues/1016). 

When `collections` is invalid, our GraphQL API errors during SDK generation with obscure messages like: `GraphQLDocumentError: Fragment cannot be spread here as objects of type "PostsAuthorDocument" can never be of type "Document"`

Reply:
We could definitely handle this in a separate ticket.